### PR TITLE
2.1.0 Release

### LINKED ISSUE #111: New Mod: EVERY
`EVERY x:` will execute the subcommand every x times.

It will be tracked **by command**, and will take 4 bytes of RAM per command, or 264 bytes total.  Additionally, it will require tracking of the current line number in the execution state, paving the way for a trampoline for SCRIPT.

If `x` changes from call to call, the behaviour will need to be sane.  I propose:

- If x_new > x, assign new mod value, no conflict.
- If x_new < x, assign new mod value, reset count to 0, produce true condition (execute the subcommand)

n.b.: I just realized that @tehn has not approved this.

### PR DESCRIPTION
**Pull request by burnsauce**
#### What does this PR do?
Updates master branch to 2.1.0

#### Provide links to any related discussion on [lines](https://llllllll.co/).
https://llllllll.co/t/teletype-2-1-firmware-beta-rc2-every-et-al/9192

#### How should this be manually tested?
Use the new features.

#### Any background context you want to provide?
All design choices and features have been approved by @tehn 

#### If the related Github issues aren't referenced in your commits, please link to them here.
EVERY implementation fixes #111 
I think the rest have been referenced by number.

#### I have,
* [ x ] updated CHANGELOG.md
* [ x ] updated the documentation

### CODE REVIEW

**Comment by samdoshi** [comment]
Just had a really quick scroll through the changes. Looks good!

I'll try to pull it locally tomorrow and make sure that everything builds cleanly on my computer.

What's the commented out code in `turtle.c` for?

**Comment by burnsauce** [comment]
Oh, those were old features of the turtle whose code I never blasted away.  I'll clean it up and push a commit.

Thanks for finding that!

**Comment by tehn** [comment]
i saw your reference to hardware 2.0 needing an RTC. not a bad idea-- i suppose we've hit the point where it'd be useful.

i'll do a bit more testing on hardware while others take time to check this all out before the merge.

**Comment by tehn** [comment]
also this is a massive amount of code-- super impressed and thank you for taking the time to be so thorough.

**Comment by scanner-darkly** [comment]
serious PR is serious :) had a look, everything seems fine. the only thing missing - could you also update https://github.com/monome/teletype/blob/master/module/help_mode.c?

thanks for doing this, can't wait to try the new features!

**Comment by burnsauce** [comment]
That file is painful to edit!

I think some reorganization is in order as I struggled to find where to put `LAST` and `BREAK`.

**Comment by burnsauce** [comment]
Uhhhh... I forgot a few changes in a branch.  Good thing I checked the network.  Sorry!  I'll try to be more careful when opening future PRs.

**Comment by samdoshi** [comment]
Another tiny little thing. You might want to rename the `sin` function in `turtle.c`, it's generating a warning:

```
../src/turtle.c:152: warning: conflicting types for built-in function 'sin'
```

---

I've checked the docs, looking good. Love the reference to "temporal recursion" in the `DEL` bug fix.

Maybe change the heading in `turtle.md` from "Turtle operator" to just "Turtle"?

> That file is painful to edit!

Isn't it just. We should be able to use the TOML files to generate the contents instead (similar to the cheat sheets). I'm pretty sure we can convert the Markdown to plain text. The only problem is that it might be too verbose and therefore take up too much space in ROM. If you're feeling super ambitious you can even build an index for it so that we can add a jump to OP documentation keybinding.

---

I need to do the school run now, will try to give it another pass this evening if I get a chance, but if not then :+1:

**Comment by cmcavoy** [comment]
I don't have anything substantive to add other than _wow this is a lot of code._ Although 2.1 is the right version number, feels like we need to include an exclamation point every time it's typed. Should be referred to as 2.1!

Super impressed.

**Comment by samdoshi** [comment]
More trivial things.

I'd be happy with `scene_number_t` going into it's own header file. (Or just do it as part of the next version.)

---

Feel free to ignore, but in `teletype.c` : `process_command`:

```c
    for (ssize_t sub_idx = 0; sub_idx < sub_len && !es_variables(es)->breaking;
         sub_idx++) {
```

Would you consider moving the `breaking` check inside the for loop?

e.g.

```c
    for (ssize_t sub_idx = 0; sub_idx < sub_len; sub_idx++) {
        if (es_variables(es)->breaking) break;  // or possible test after the op/mod has run
```

Don't make the change now though, but maybe for the next release.

This is probably the non-C code in me talking, but I do find some of those C-isms to be non-intuitive. (Also see assignment in conditionals).

----

Regarding `ss_clear_script`, can you not just clear a script by setting it's length to 0?

From `state.c`:

```c
void ss_clear_script(scene_state_t *ss, size_t script_idx) {
    for (uint8_t i = 0; i < SCRIPT_MAX_COMMANDS; i++)
        if (ss_get_script_len(ss, script_idx))
            ss_delete_script_command(ss, script_idx, i);
}
```

becomes:

```c
void ss_clear_script(scene_state_t *ss, size_t script_idx) {
        ss_set_script_len(ss, script_idx, 0);
}
```

`ss_delete_script_command` is not going to be the fastest (this is used with TEMP scripts a lot, right?). Again, probably best for a later version as the change will need testing.

----

Anyway, thank you _so_ much for all of this. It looks really solid.

Are you ready for it to be merged?

**Comment by burnsauce** [comment]
Re: ss_clear_script()

Uh... I must have been out of it when I wrote that code.  It doesn't even make sense.  I'll definitely rewrite it, and will probably use your suggestion.

Good job spotting that, and thank you for taking the time to review the code and find it!

One more commit to clear that up, and it'll be ready for merge.

**Comment by burnsauce** [comment]
If I had to guess, I wrote the if and the for statement in the wrong order or it was a copy-paste mistake.  Maybe my caps-lock was on, and I typed P instead of p in vim.

Anyhow, you were right, @samdoshi, all I need to do is set the script length to zero.

OK to merge IMO!  I agree, let's postpone other changes to the next release.

**Comment by burnsauce** [comment]
After it's merged, I guess @tehn will have to build the release for hosting.  Despite downloading a fresh texlive install, I still get too-wide tables, and I don't have the time to further diagnosis.

**Comment by tehn** [comment]
certainly, i'll post up the release with proper txt and all.

unfortunately i'm still not set up to build the pdf docs, if someone could
upload that to the forum.

i'll get the docs page updated on the main site as well.

are we feeling final? shall i merge?

**Comment by burnsauce** [comment]
Yes, I'm done making changes to this branch.  If bugs get discovered past this point, that's what the third digit in the release number is for! ;)

Merge away!

**Comment by samdoshi** [comment]
@tehn did this get "squash merged"?

**Comment by tehn** [comment]
yes, but then i had to do a couple master push for doc updates.

**Comment by samdoshi** [comment]
Did you mean to "squash" it? (i.e. turn the >100 commits into a single one)

Be careful, GitHub changed the merge box on PRs to give 3 options.

**Comment by tehn** [comment]
indeed, that was the default (which i did notice was different).

now that you're bringing this up, of course it makes sense that i should've
left all of the commits atomic for traceability, right?

**Comment by samdoshi** [comment]
Usually yes. Sometimes you want to use the other options.

Also, it's messed up the contributors graph a bit, so @burnsauce doesn't get quite as much recognition on it as he deserves. (Though we all know how much work he has put into it!)

**Comment by burnsauce** [comment]
I must say that I was looking forward to seeing my 200+ commits appear on the graph.

It's hard for me to put forward technical arguments as to why it shouldn't have been squashed because they would be disingenuous in light of that.

Just posting this here to voice my opinion before too many more commits stack up and it might be harder to redo the PR.

**Comment by scanner-darkly** [comment]
i think recognizing everybody's contributions is super important. but feels like commit history should be more driven by what makes it more useful/efficient for us. i had commits that were some silly typo or such, i wouldn't want them creating noise on the master branch. but yeah i'm not sure how to balance it with trackability... is there a way to cherry pick what to squash? interesting question this...

**Comment by samdoshi** [comment]
So firstly, if this had happened to me, I would feel pretty hacked off. By all metrics @burnsauce  should be the 2nd highest contributor to the repo behind me.

That commit history will also be really helpful with `blame` and `bisect`. _Really helpful._

The only way to solve this now is to do some history rewriting and then force pushing.

I'm pretty sure that I can do that, by branching of `burnsauce/teletype` and cherry picking the newer commits on top of it, **but**, the hardest bit will be for @scanner-darkly to do. He's already rebased his Teletype OPs branch on top of this PR.

So basically, we're stuck between a rock and a hard place. Whatever we do, I think we should decide quickly.

**Comment by samdoshi** [comment]
Right one more option, the following branch:

https://github.com/samdoshi/teletype/tree/burnsauce-master

Has all of @burnsauce commits merged back in, because they're merged in there is no need to force push. But it does make the commit history look really really wonky.

Here is the diff:

https://github.com/monome/teletype/compare/master...samdoshi:burnsauce-master

You can see all the commits, but also see that it doesn't change anything.

Here is the blame:

https://github.com/samdoshi/teletype/blame/burnsauce-master/src/turtle.c

You can see that all the original commits are showing up.

**Comment by burnsauce** [comment]
@samdoshi Thank you for following up on this!

I really appreciate the work and attention that you put into this, especially considering your limited time.

Also I apologise for claiming it's 200+ commits, as it's only 100+.  I think I might be developing dyslexia of some sort.  Numeric transposition has been happening to me with alarming regularity.

**Comment by tehn** [comment]
augh, i am terribly sorry for this. there was no technical reason, other
than my over-multi-tasking caused me to do the default merge (which seems
to have changed in the github UI?).

i am up for any solution that can fix this and bestow the proper merit on
@burnsauce, to whom i'm super grateful. again i apologize for the mess.

**Comment by samdoshi** [comment]
@tehn yeah GitHub changed the UI, the default option seems to change, so you really need to focus on it. It's good that they offer the options now, but it is confusing.

Shall I open up a PR to merge the commits back in then? I think it works, but it's a bit at the edge of my `git` knowledge. @cmcavoy  / @jlmitch5 either of you have any insight?

**Comment by tehn** [comment]
please do, if you have the time to attempt to rectify this. thank you!

**Comment by burnsauce** [comment]
Thank you both for taking the time!  Sorry to @scanner-darkly for any additional work this will bestow upon your grid fork :(

**Comment by jlmitch5** [comment]
Agreed the expanded commit history is very useful (especially for someone, like myself, using this repo as a way to learn more about C/C++ embedded dev).

I don't really understand what you did with that PR @samdoshi, but it looks to be working great!  I would say this way is especially good because it might be weird to rewrite history with force pushes since the master branch has releases off of it.  I don't know what would happen to them.

That brings up a point...I guess this would be merged into master, so it wouldn't be present in the 2.1.0 release tag, but would be for all subsequent releases off of master.  I assume that's fine just wanted to bring it up.

**Comment by burnsauce** [comment]
Somewhat painful news:  the prune job I did removing the .zip files and .hex files that I once added to the repo did not take.

I only found this out by checking burnsauce-master@samdoshi/teletype with BFG which means that I did push those things onto the branch.

Seems that history will need to be rewritten at least for 2.1.0.  Sorry that my git-newbieness has dirtied the repo.

For reference, I used https://rtyley.github.io/bfg-repo-cleaner/ with
```
java -jar bfg.jar -D '*.zip'
git reflog expire --expire=now --all && git gc --prune=now --aggressive
java -jar bfg.jar -D '*.hex'
git reflog expire --expire=now --all && git gc --prune=now --aggressive
```
and it found stuff that I thought I had already deleted with `git filter-branch`.

:(

**Comment by samdoshi** [comment]
@burnsauce get a cleaned up branch sorted for me, then I can take it from there.

Because @tehn squash merged it, none of the zip files are in the current `master`. If you can get a cleaned up branch, I can redo the merge and open up a PR. If it's too difficult to clean up your branch, then I suggest we keep `master` as is, and chalk this one up to experience.

**Comment by burnsauce** [comment]
My master has been cleaned of these references.

```
Deleted files
-------------

        Filename                 Git id                                                       
        --------------------------------------------------------------------------------------
        teletype-2.1-beta1.zip | 718fc9e8 (414.8 KB), 74e18c23 (414.8 KB)                     
        teletype-2.1-beta2.zip | 2d940c8c (415.2 KB)                                          
        teletype-2.1-beta3.zip | 6c674ff8 (411.8 KB)                                          
        teletype-2.1-beta4.zip | c42995be (412.3 KB)                                          
        teletype-2.1-beta5.zip | 1fe08931 (412.5 KB), 274b0a97 (412.2 KB), fb8421c7 (412.3 KB)

```

```
Deleted files
-------------

        Filename       Git id                                                       
        ----------------------------------------------------------------------------
        teletype.hex | c341ec91 (345.6 KB), 1f9f47f0 (331.5 KB), 3f9f4909 (346.2 KB)

```

**Comment by samdoshi** [comment]
Can you check for PDF files too?

**Comment by burnsauce** [comment]
No PDF files found! :)

**Comment by scanner-darkly** [comment]
for my grid2 branch i could probably just rebase off whatever the master ends up with? this shouldn't be a problem if i understand correctly, it should be able to replay my changes with no conflicts. 

and i think we should have a discussion on commit etiquette - blame and tracking is very useful, but maybe that means putting more care into what goes into a commit - when i start development i might have a bunch of commits that is work in progress/just me forgetting to do things/silly typos, i don't think those would be useful for blame and would just create more noise for others. thoughts?

**Comment by burnsauce** [comment]
If there were an easy way to selectively roll up a sequential series of commits into one after the fact, that would allow one to hide typo and other premature commit fixes while retaining the richness of the commit metadata.

Is that something git can do?

**Comment by samdoshi** [comment]
Yes, you need to Google "git interactive rebase". Read through a couple of the links to get an idea of how it works.

One thing to remember with `git`, is that even when rewriting history, old commits are never lost (not until the garbage collector runs, and that is pretty infrequent), but they can be hard to easily find again. So whenever I'm about to do something tricky with history in `git`, I always like to make a branch (e.g. `git branch pre-rebase`), so that I can always get back to where I started when it all goes horribly wrong!

(I've had a busy day, so I'm not going to be able to doing any more on this today. So if you want to use the time to rewrite history a bit, go for it.)

**Comment by jlmitch5** [comment]
Interactive rebase/squashing is a good option.

A couple other notes (sorry if this is all super obvious):
- It is good practice to `get pull --rebase` off the branch you're basing off of.  This keeps your changes on top of the current state, even if there are commits in the branch getting pulled that are more recent than some of your commits.  This keeps your changes as a section that can cleanly go back in the tree.
- soft reset can be really useful.  `get reset head~n` will unstage the n top-most commits.  You can then commit pieces of the unstaged changes in a more semantic way (use `git add -p` to run through the unstaged changes interactively).  I typically find this easier to work with that interactive rebasing, but I think it depends on the type of changes...this would probably be pretty difficult with a single PR of a very large amount of commits.

> I always like to make a branch (e.g. git branch pre-rebase), so that I can always get back to where I started when it all goes horribly wrong!

This is very good advice, and really, any time you are about to do `git push --force` to rewrite a remote it's a good idea to have the changes somewhere so you don't accidentally lose them.

**Comment by burnsauce** [comment]
I gave rewriting the history for my changes a shot, but the squashes are less automatic than I expected, leading me to have to manually fix merges.  As this could lead to bugs, I'm not going to engage in this activity at this point in time.

Knowing how to squash _will_ lead me to cleaner commit logs in the future.  I have no interest in creating noise, and would consider a feature/bugfix PLUS its documentation 1 commit.

@samdoshi, you can consider my master branch as ready to go whenever you want to commit surgery.  Thank you all for the guidance.
